modules/gltf/gltf_document.cpp

```diff
@@ -2241,118 +2235,188 @@ Error GLTFDocument::_serialize_meshes(Ref<GLTFState> state) {
+						if (total > 0.0) {
+							weight_0.r /= total;
+							weight_0.g /= total;
+							weight_0.b /= total;
+							weight_0.a /= total;
 						}
 					}
-					primitive["indices"] = _encode_accessor_as_ints(state, mesh_indices, true);
-				} else {
-					if (primitive_type == Mesh::PRIMITIVE_TRIANGLES) {
-						//generate indices because they need to be swapped for CW/CCW
-						const Vector<Vector3> &vertices = array[Mesh::ARRAY_VERTEX];
-						Ref<SurfaceTool> st;
-						st.instance();
-						st->create_from_triangle_arrays(array);
-						st->index();
-						Vector<int32_t> generated_indices = st->commit_to_arrays()[Mesh::ARRAY_INDEX];
-						const int vs = vertices.size();
-						generated_indices.resize(vs);
-						{
-							for (int k = 0; k < vs; k += 3) {
-								generated_indices.write[k] = k;
-								generated_indices.write[k + 1] = k + 2;
-								generated_indices.write[k + 2] = k + 1;
-							}
+					Color joint_0;
+					joint_0.r = 0;
+					joint_0.g = 0;
+					joint_0.b = 0;
+					joint_0.a = 0;
+					Set<BoneId> set_bones;
+					_normalize_bone(array_bones[(i * JOINT_GROUP_SIZE) + 0], set_bones, joint_0.r, weight_0.r);
+					_normalize_bone(array_bones[(i * JOINT_GROUP_SIZE) + 1], set_bones, joint_0.g, weight_0.g);
+					_normalize_bone(array_bones[(i * JOINT_GROUP_SIZE) + 2], set_bones, joint_0.b, weight_0.b);
+					_normalize_bone(array_bones[(i * JOINT_GROUP_SIZE) + 3], set_bones, joint_0.a, weight_0.a);
+					joint_attribs.write[i] = joint_0;
+					{
+						float total = 0.0;
+						total += weight_0.r;
+						total += weight_0.g;
+						total += weight_0.b;
+						total += weight_0.a;
+						if (total > 0.0) {
+							weight_0.r /= total;
+							weight_0.g /= total;
+							weight_0.b /= total;
+							weight_0.a /= total;
 						}
-						primitive["indices"] = _encode_accessor_as_ints(state, generated_indices, true);
+						weight_attribs.write[i] = weight_0;
 					}
 				}
+				attributes["WEIGHTS_0"] = _encode_accessor_as_weights(state, weight_attribs, true);
+				attributes["JOINTS_0"] = _encode_accessor_as_joints(state, joint_attribs, true);
 			}
 
+			Vector<int32_t> mesh_indices = array[Mesh::ARRAY_INDEX];
+			if (mesh_indices.size()) {
+				if (primitive_type == Mesh::PRIMITIVE_TRIANGLES) {
+					//swap around indices, convert ccw to cw for front face
+					const int is = mesh_indices.size();
+					for (int k = 0; k < is; k += 3) {
+						SWAP(mesh_indices.write[k + 0], mesh_indices.write[k + 2]);
+					}
+				}
+				primitive["indices"] = _encode_accessor_as_ints(state, mesh_indices, true);
+			} else {
+				if (primitive_type == Mesh::PRIMITIVE_TRIANGLES) {
+					//generate indices because they need to be swapped for CW/CCW
+					const Vector<Vector3> &vertices = array[Mesh::ARRAY_VERTEX];
+					Ref<SurfaceTool> st;
+					st.instance();
+					st->create_from_triangle_arrays(array);
+					st->index();
+					Vector<int32_t> generated_indices = st->commit_to_arrays()[Mesh::ARRAY_INDEX];
```

Comment:
thats not common practice. store the arrays in an own variable.
check if index are exist in the array and it has items